crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::circuit_instruction::CircuitInstruction;
+use pyo3::prelude::*;
+use pyo3::sync::GILOnceCell;
+use pyo3::types::{PyDict, PyList, PySequence, PyString, PyTuple};
+use pyo3::{intern, PyObject, PyResult};
+
+static SEMANTIC_EQ_PYTHON: GILOnceCell<PyResult<PyObject>> = GILOnceCell::new();
+
+/// Parent class for DAGOpNode, DAGInNode, and DAGOutNode.
+#[pyclass(module = "qiskit._accelerate.circuit", subclass)]
+#[derive(Clone, Debug)]
+pub struct DAGNode {
+    #[pyo3(get, set)]
+    pub _node_id: isize,
+}
+
+#[pymethods]
+impl DAGNode {
+    #[new]
+    #[pyo3(signature=(nid=-1))]
+    fn new(nid: isize) -> Self {
+        DAGNode { _node_id: nid }
+    }
+
+    fn __getstate__(&self) -> isize {
+        self._node_id
+    }
+
+    fn __setstate__(&mut self, nid: isize) {
+        self._node_id = nid;
+    }
```

Comment:
Rather than setting `nid = -1`, can we make this `Option<T>` if we _need_ it to be undefined, with a view to removing the `Option` once this moves properly to Rust space?

We also don't need `__getstate__` and `__setstate__` here if we define `__getnewargs__` to return `(self._node_id,)`, which will make copy/pickle a bit more efficient.

Reply:
Are you suggesting that `DAGNode` store an `Option<petgraph::NodeIndex>` instead of an `isize`?

The reason I didn't attempt to do this for now is because technically `-1` as the default is part of the public API for `DAGNode` and all subclasses. What'd be your preference here?

Reply:
My main concern was to avoid encoding the `isize` into Rust space, since Rust has much better ways to represent that in safe typing.  That said, longer term, I also feel like we won't want to be storing the `_node_id` inside the weights of our graph object at all, because the graph weights won't be accessible from Python space without first getting unpacked and wrapped into Python objects from Rust space, at which point we could manually push a valid `_node_id` in.

So really, I guess that means I don't care too much about `DAGNode` here, because it oughtn't to stay our internal Rust data anyway.

Reply:
When we're at the point of doing that, we can make this a proper `NodeIndex` or just `u32` for exposure to Python space, since Python space won't be constructing these things manually anyway (with suitable deprecation warnings ahead of 2.0).